src/interface/sparse_pack.hpp

```diff
@@ -0,0 +1,282 @@
+// (C) (or copyright) 2020-2022. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#ifndef INTERFACE_SPARSE_PACK_HPP_
+#define INTERFACE_SPARSE_PACK_HPP_
+
+#include <algorithm>
+#include <functional>
+#include <limits>
+#include <map>
+#include <memory>
+#include <regex>
+#include <string>
+#include <tuple>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include "coordinates/coordinates.hpp"
+#include "interface/sparse_pack_base.hpp"
+#include "interface/variable.hpp"
+#include "utils/utils.hpp"
+
+namespace parthenon {
+
+namespace impl {
+template <int... IN>
+struct multiply;
+
+template <>
+struct multiply<> : std::integral_constant<std::size_t, 1> {};
+
+template <int I0, int... IN>
+struct multiply<I0, IN...> : std::integral_constant<int, I0 * multiply<IN...>::value> {};
+
+// GetTypeIdx is taken from Stack Overflow 26169198, should cause compile time failure if
+// type is not in list
+template <typename T, typename... Ts>
+struct GetTypeIdx;
+
+template <typename T, typename... Ts>
+struct GetTypeIdx<T, T, Ts...> : std::integral_constant<std::size_t, 0> {
+  using type = void;
+};
+
+template <typename T, typename U, typename... Ts>
+struct GetTypeIdx<T, U, Ts...>
+    : std::integral_constant<std::size_t, 1 + GetTypeIdx<T, Ts...>::value> {
+  using type = void;
+};
+
+template <class T, class... Ts>
+struct IncludesType;
+
+template <typename T>
+struct IncludesType<T, T> : std::true_type {};
+
+template <typename T, typename... Ts>
+struct IncludesType<T, T, Ts...> : std::true_type {};
+
+template <typename T, typename U>
+struct IncludesType<T, U> : std::false_type {};
+
+template <typename T, typename U, typename... Ts>
+struct IncludesType<T, U, Ts...> : IncludesType<T, Ts...> {};
+
+} // namespace impl
+
+using namespace impl;
+
+namespace variables {
+// Struct that all variables types should inherit from
+template <bool REGEX, int... NCOMP>
+struct base_t {
+  KOKKOS_INLINE_FUNCTION
+  base_t() : idx(0) {}
+
+  KOKKOS_INLINE_FUNCTION
+  explicit base_t(int idx1) : idx(idx1) {}
+
+  virtual ~base_t() = default;
+
+  // All of these are just static methods so that there is no
+  // extra storage in the struct
```

Comment:
Should these be marked `virtual` too to avoid shadowing warnings? Or does that just happen automatically with `static`?

Reply:
I think it just happens automatically and I don't think there is any virtual look up table for static functions. I have not seen any shadowing warnings.